gtsam/nonlinear/GncOptimizer.h

```diff
@@ -0,0 +1,425 @@
+    for (size_t i = 0; i < nfg_.size(); i++) {
+      if (nfg_[i]) {
+        rmax_sq = std::max(rmax_sq, nfg_[i]->error(state_));
+      }
+    }
+    // set initial mu
+    switch (params_.lossType) {
+    case GncParameters::GM:
+      // surrogate cost is convex for large mu
+      return 2 * rmax_sq / params_.barcSq; // initial mu
+    case GncParameters::TLS:
+      // initialize mu to the value specified in Remark 5 in GNC paper.
+      // surrogate cost is convex for mu close to zero
+      // degenerate case: 2 * rmax_sq - params_.barcSq < 0 (handled in the main loop)
+      return params_.barcSq / (2 * rmax_sq - params_.barcSq);
+    default:
+      throw std::runtime_error(
+          "GncOptimizer::initializeMu: called with unknown loss type.");
+    }
+  }
+
+  /// update the gnc parameter mu to gradually increase nonconvexity
+  double updateMu(const double mu) const {
+    switch (params_.lossType) {
+    case GncParameters::GM:
+      // reduce mu, but saturate at 1 (original cost is recovered for mu -> 1)
+      return std::max(1.0, mu / params_.muStep);
+    case GncParameters::TLS:
+      // increases mu at each iteration (original cost is recovered for mu -> inf)
+      return mu * params_.muStep;
+    default:
+      throw std::runtime_error(
+          "GncOptimizer::updateMu: called with unknown loss type.");
+    }
+  }
+
+  /// calculated sum of weighted squared residuals
+  double calculateWeightedCost() const {
+    double cost = 0;
+    for (size_t i = 0; i < nfg_.size(); i++) {
+      cost += weights_[i] * nfg_[i]->error(state_);
+    }
+    return cost;
+  }
+
+  /// check if we have reached the value of mu for which the surrogate loss matches the original loss
+  bool checkMuConvergence(const double mu) const {
+    switch (params_.lossType) {
+    case GncParameters::GM:
+      return std::fabs(mu - 1.0) < 1e-9; // mu=1 recovers the original GM function
+    default:
+      throw std::runtime_error(
+          "GncOptimizer::checkMuConvergence: called with unknown loss type.");
+    }
+  }
+
+  /// check convergence of relative cost differences
+  bool checkCostConvergence(const double cost, const double prev_cost) const {
+    switch (params_.lossType) {
+    case GncParameters::TLS:
+      return std::fabs(cost - prev_cost) < params_.relativeCostTol;
+    default:
+      throw std::runtime_error(
+          "GncOptimizer::checkMuConvergence: called with unknown loss type.");
+    }
+  }
+
+  /// check for convergence between consecutive GNC iterations
+  bool checkConvergence(const double mu,
+                        const double cost,
+                        const double prev_cost) const {
+    switch (params_.lossType) {
+    case GncParameters::GM:
+      return checkMuConvergence(mu);
+    case GncParameters::TLS:
+      return checkCostConvergence(cost, prev_cost);
+    default:
+      throw std::runtime_error(
+          "GncOptimizer::checkMuConvergence: called with unknown loss type.");
+    }
+  }
```

Comment:
This will work well, but it looks weird that we use two very different stopping criteria for the two robust costs. I would use the cost based convergence test in both cases as default and add a parameter to support three convergence tests:
1. cost based only (default)
2. weighs based only
3. [optional] cost based OR weights base (whatever converges first)
I think this would add clarity overall.

Reply:
for simplicity, I would add both stopping conditions (for both costs) and give the user control of the tolerances for stopping, such that s/he can enable/disable the conditions s/he prefers.

Reply:
@jingnanshi : let me know if you can take a look at this, otherwise I can do it later in December.

Reply:
@lucacarlone @pantonante For the TLS case, since mu is increasing, it is difficult to check for mu (unless I have misunderstood it). What do you guys recommend?

Reply:
@jingnanshi : Pasquale was suggesting two checks I think:
1) relative change in cost: e.g., | cost_t - cost_t-1 | /  |cost_t | < params_.relativeCostTol (then stop)
2) weights: e.g., if | weight - round(weights) | < params_.weightConvergenceTol  (then stop)
In the latter round(weights) rounds the entries of weights in {0;1} and this is mostly effective for TLS (it checks whether you have a clear decision inlier/outliers).

Note: you can also keep the stopping condition based on mu for GM, while there is no such condition for TLS